src/proposer/proposer_logic.cpp

```diff
@@ -49,27 +49,36 @@ class LogicImpl final : public Logic {
     if (!current_tip) {
       return boost::none;
     }
+
     const blockchain::Height current_height =
         m_active_chain->GetHeight();
     const blockchain::Height target_height =
         current_height + 1;
+
+    int64_t best_time = std::max(current_tip->GetMedianTimePast() + 1, m_network->GetTime());
     const blockchain::Time target_time =
-        m_blockchain_behavior->CalculateProposingTimestampAfter(m_network->GetTime());
+        m_blockchain_behavior->CalculateProposingTimestampAfter(best_time);
     const blockchain::Difficulty target_difficulty =
         m_blockchain_behavior->CalculateDifficulty(target_height, *m_active_chain);
 
     for (const staking::Coin &coin : eligible_coins) {
       const uint256 kernel_hash = m_stake_validator->ComputeKernelHash(current_tip, coin, target_time);
-      if (m_stake_validator->CheckKernel(coin.GetAmount(), kernel_hash, target_difficulty)) {
-        const CAmount reward = m_blockchain_behavior->CalculateBlockReward(
-            target_height);
-        return {{coin,
-                 kernel_hash,
-                 reward,
-                 target_height,
-                 target_time,
-                 target_difficulty}};
+
+      if (!m_stake_validator->CheckKernel(coin.GetAmount(), kernel_hash, target_difficulty)) {
+        if (m_blockchain_behavior->GetParameters().mine_blocks_on_demand) {
+          LogPrint(BCLog::VALIDATION, "Letting artificial block generation succeed nevertheless (mine_blocks_on_demand=true)\n");
```

Comment:
The category here should not be `VALIDATION` but `PROPOSING`:

```
    {BCLog::PROPOSING, "proposing"},
```

As we are in `src/proposer` and the `proposer` namespace, not the `staking` namespace which is part of validation logic.